pkg/model/firewall.go

```diff
@@ -42,72 +42,74 @@ type FirewallModelBuilder struct {
 
 func (b *FirewallModelBuilder) Build(c *fi.ModelBuilderContext) error {
-	if err := b.buildNodeRules(c); err != nil {
+	nodeGroups, err := b.buildNodeRules(c)
+	if err != nil {
 		return err
 	}
-	if err := b.buildMasterRules(c); err != nil {
+
+	masterGroups, err := b.buildMasterRules(c, nodeGroups)
+	if err != nil {
 		return err
 	}
+
+	// We _should_ block per port... but:
+	// * It causes e2e tests to break
+	// * Users expect to be able to reach pods
+	// * If users are running an overlay, we punch a hole in it anyway
+	// re-using provided security group for the node
+	b.applyNodeToMasterBlockSpecificPorts(c, nodeGroups, masterGroups)
+
 	return nil
 }
 
-func (b *FirewallModelBuilder) buildNodeRules(c *fi.ModelBuilderContext) error {
-	{
-		t := &awstasks.SecurityGroup{
-			Name:             s(b.SecurityGroupName(kops.InstanceGroupRoleNode)),
-			Lifecycle:        b.Lifecycle,
-			VPC:              b.LinkToVPC(),
-			Description:      s("Security group for nodes"),
-			RemoveExtraRules: []string{"port=22"},
-		}
-		c.AddTask(t)
-	}
+func (b *FirewallModelBuilder) buildNodeRules(c *fi.ModelBuilderContext) (map[string]*awstasks.SecurityGroup, error) {
 
-	// Allow full egress
-	{
-		t := &awstasks.SecurityGroupRule{
-			Name:          s("node-egress"),
-			Lifecycle:     b.Lifecycle,
-			SecurityGroup: b.LinkToSecurityGroup(kops.InstanceGroupRoleNode),
-			Egress:        fi.Bool(true),
-			CIDR:          s("0.0.0.0/0"),
-		}
-		c.AddTask(t)
+	nodeGroups, err := CreateSecurityGroups(b.InstanceGroups, kops.InstanceGroupRoleNode, b.ClusterName(), b.Lifecycle, b.LinkToVPC(), c)
+	if err != nil {
+		return nil, err
 	}
 
-	// Nodes can talk to nodes
-	{
-		t := &awstasks.SecurityGroupRule{
-			Name:          s("all-node-to-node"),
-			Lifecycle:     b.Lifecycle,
-			SecurityGroup: b.LinkToSecurityGroup(kops.InstanceGroupRoleNode),
-			SourceGroup:   b.LinkToSecurityGroup(kops.InstanceGroupRoleNode),
+	for nodeGroupName, secGroup := range nodeGroups {
+		suffix := GetGroupSuffix(nodeGroupName, nodeGroups)
+		// Allow full egress
+		{
+			t := &awstasks.SecurityGroupRule{
+				Name:          s(fmt.Sprintf("node-egress%s", suffix)),
+				Lifecycle:     b.Lifecycle,
+				SecurityGroup: secGroup,
+				Egress:        fi.Bool(true),
+				CIDR:          s("0.0.0.0/0"),
+			}
+			c.AddTask(t)
 		}
-		c.AddTask(t)
-	}
 
-	// Pods running in Nodes could need to reach pods in master/s
-	if b.Cluster.Spec.Networking != nil && b.Cluster.Spec.Networking.AmazonVPC != nil {
-		// Nodes can talk to masters
+		// Nodes can talk to nodes
 		{
 			t := &awstasks.SecurityGroupRule{
-				Name:          s("all-nodes-to-master"),
+				Name:          s(fmt.Sprintf("all-node-to-node%s", suffix)),
 				Lifecycle:     b.Lifecycle,
-				SecurityGroup: b.LinkToSecurityGroup(kops.InstanceGroupRoleMaster),
-				SourceGroup:   b.LinkToSecurityGroup(kops.InstanceGroupRoleNode),
+				SecurityGroup: secGroup,
+				SourceGroup:   secGroup,
```

Comment:
I think this just allows each node SG to talk to itself, but we want the full cross-product